### PR TITLE
[c++] Fix memory leak in Arrow table implementation

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

When trying to integrate LightGBM v4.2 into our codebase, we noticed an unexpected increase in memory consumption when using LightGBM's new Arrow interface under certain circumstances.

It turns out that the Arrow implementation causes memory leaks as it does not properly release the Arrow arrays being passed to it. Unfortunately, this issue only really surfaces if the data being passed to Arrow is not in memory during the entire program execution (which is frequently the case) and the leakage really only becomes an issue if different data is passed to the LightGBM Arrow interface multiple times (this is where we discovered it).

The fix is simple and the reasoning behind it is outlined clearly in the Arrow C Data interface [specification](https://arrow.apache.org/docs/format/CDataInterface.html#memory-management).

---

Huge shoutout to @delsner who helped me debug this memory leak and helped me learn a lot more about Arrow, allocators and memory debugging 😁

---

Sorry for not already including this in #6034 and its sibling PRs... I still had some gaps in my understanding of the Arrow C Data interface :eyes:

### CODE REVIEW

**Comment by borchero** [comment]
FYI, if anyone has a good idea how to write good tests for this, please let me know :eyes: once this is merged, we should probably also cut a new release soonish to allow people to confidently use the Arrow interface.

File: include/LightGBM/arrow.h
Comment (jameslamb) [inline include/LightGBM/arrow.h:239]: thanks for including this link, super helpful!

**Review by jameslamb — COMMENTED** [review]
Thanks! I'll defer to @guolinke or @shiyu1994 for a review on this.

Just want to add that if/when we add support for Arrow data via the R-package here, we should be able to catch such things through the `valgrind` checks we run on the R package: https://github.com/microsoft/LightGBM/tree/master/R-package#valgrind

**Comment by borchero** [comment]
> Just want to add that if/when we add support for Arrow data via the R-package here

Would be great! I myself, however, have close to no experience with R, unfortunately 😅

**Review by shiyu1994 — APPROVED** [review]
Thanks for the excellent fix.

**Comment by borchero** [comment]
@jameslamb can you tell me how to retry jobs run on Azure? I can't seem to find a button that allows me to do this.

**Comment by shiyu1994** [comment]
@borchero Could you register your an Azure DevOps account with the same email as your Github?

**Comment by borchero** [comment]
Done @shiyu1994, thanks a lot! :)

**Pull request merged**
